apollo-federation/src/correctness/response_shape_compare.rs

```diff
@@ -0,0 +1,387 @@
+        if let Ok(other_def) = collect_definitions_for_type_condition(other, this_cond) {
+            let result = compare_possible_definitions_per_type_condition(
+                &updated_constraint,
+                this_def,
+                &other_def,
+            );
+            match result {
+                Ok(result) => return Ok(result),
+                Err(err) => {
+                    // See if we can case-split over ground set items.
+                    if this_cond.ground_set().len() == 1 {
+                        // Single object type has no other option. Stop and report the error.
+                        let detail = detail_single_object_type_condition(this_cond);
+                        return Err(CheckFailure::new(format!(
+                            "mismatch for type condition: {this_cond}{detail}\n{}",
+                            err.description()
+                        )));
+                    }
+                }
+            }
+            // fall through
+        };
+
+        // Finally: Case-split over individual ground types.
+        let ground_set_iter = this_cond.ground_set().iter();
+        let mut ground_set_iter = ground_set_iter.filter(|ty| path_constraint.allows(ty));
+        ground_set_iter.try_for_each(|ground_ty| {
+            let filter_cond = NormalizedTypeCondition::from_object_type(ground_ty);
+            let other_def =
+                collect_definitions_for_type_condition(other, &filter_cond).map_err(|e| {
+                    e.add_description(&format!(
+                        "missing type condition: {this_cond} (case: {ground_ty})"
+                    ))
+                })?;
+            let updated_constraint = path_constraint.under_type_condition(&filter_cond);
+            compare_possible_definitions_per_type_condition(
+                &updated_constraint,
+                this_def,
+                &other_def,
+            )
+            .map_err(|e| {
+                e.add_description(&format!(
+                    "mismatch for type condition: {this_cond} (case: {ground_ty})"
+                ))
+            })
+        })
+    })
+}
+
+fn compare_possible_definitions_per_type_condition<'a, T: PathConstraint<'a>>(
+    path_constraint: &T,
+    this: &PossibleDefinitionsPerTypeCondition,
+    other: &PossibleDefinitionsPerTypeCondition,
+) -> Result<(), CheckFailure> {
+    compare_field_selection_key(this.field_selection_key(), other.field_selection_key()).map_err(
+        |e| {
+            e.add_description(
+                "mismatch in field selection key of PossibleDefinitionsPerTypeCondition",
+            )
+        },
+    )?;
+    this.conditional_variants()
+        .iter()
+        .try_for_each(|this_def| {
+            // search the same boolean clause in other
+            let found = other
+                .conditional_variants()
+                .iter()
+                .fallible_any(|other_def| {
```

Comment:
I think the `fallible_find` method was removed during cleanup, but this would be an excellent place to bring it back.